editor/src/plugins/tilemap/brush_macro.rs

```diff
@@ -0,0 +1,992 @@
+        &mut self,
+        prop: Option<&TileSetPropertyLayer>,
+        message: &UiMessage,
+        ui: &mut UserInterface,
+    ) {
+        if let Some(TextMessage::Text(text)) = message.data() {
+            if message.destination() == self.textbox
+                && message.direction() == MessageDirection::FromWidget
+            {
+                ui.send_message(TileSetPropertyValueMessage::value(
+                    self.handle,
+                    MessageDirection::FromWidget,
+                    TileSetPropertyValueElement::String(text.into()),
+                ));
+            }
+        } else if let Some(NumericUpDownMessage::<i8>::Value(v)) = message.data() {
+            if message.destination() == self.textbox
+                && message.direction() == MessageDirection::FromWidget
+            {
+                ui.send_message(TileSetPropertyValueMessage::value(
+                    self.handle,
+                    MessageDirection::FromWidget,
+                    TileSetPropertyValueElement::I8(*v),
+                ));
+                send_sync_message(
+                    ui,
+                    DropdownListMessage::selection(
+                        self.list,
+                        MessageDirection::ToWidget,
+                        Some(find_list_index(prop, NamableValue::I8(*v))),
+                    ),
+                );
+            }
+        } else if let Some(NumericUpDownMessage::<i32>::Value(v)) = message.data() {
+            if message.destination() == self.textbox
+                && message.direction() == MessageDirection::FromWidget
+            {
+                ui.send_message(TileSetPropertyValueMessage::value(
+                    self.handle,
+                    MessageDirection::FromWidget,
+                    TileSetPropertyValueElement::I32(*v),
+                ));
+                send_sync_message(
+                    ui,
+                    DropdownListMessage::selection(
+                        self.list,
+                        MessageDirection::ToWidget,
+                        Some(find_list_index(prop, NamableValue::I32(*v))),
+                    ),
+                );
+            }
+        } else if let Some(NumericUpDownMessage::<f32>::Value(v)) = message.data() {
+            if message.destination() == self.textbox
+                && message.direction() == MessageDirection::FromWidget
+            {
+                ui.send_message(TileSetPropertyValueMessage::value(
+                    self.handle,
+                    MessageDirection::FromWidget,
+                    TileSetPropertyValueElement::F32(*v),
+                ));
+                send_sync_message(
+                    ui,
+                    DropdownListMessage::selection(
+                        self.list,
+                        MessageDirection::ToWidget,
+                        Some(find_list_index(prop, NamableValue::F32(*v))),
+                    ),
+                );

```

Comment:
These three blocks looks almost identical and their difference is `TileSetPropertyValueElement` and `NamableValue` only, I think there should be a function that accepts such args and replace these three blocks with this function call.